---
title: "Mime-16s-emp"
author: "Clara Jégousse"
date: "3/3/2021"
output: html_document
---

# Load libraries

```{r, echo=FALSE}
library(devtools)
library(ggplot2)
library(ggpubr)
library(dada2)
library(phyloseq)
library(reshape2) # to use melt
```

# Visual setting

Import variables and functions for consistent plots.

```{r}
source_url("https://raw.githubusercontent.com/clarajegousse/mime-16s/main/scripts/visual-settings.r")
```

# Read Phyloseq Object

Import the phyloseq object containing all 1397 samples amplified with the EMP primers with metadata from Hafro. 

```{r, echo=FALSE}
ps <- readRDS("/Users/Clara/Projects/mime-16s/global-ps-emp.rds")

dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps

```

All metadata including measures from Hafro.
```{r}
sample_variables(ps)
```

```{r}
rank_names(ps)
```

Number of taxa
```{r}
ntaxa(ps)
```

# Deal with missing values in metadata with imputation of the mean of the two surrounding values

```{r}
library(imputeTS)
sample_data(ps)$po4.avg <- round(na_ma(sample_data(ps)$po4.avg, k = 1), digits = 2)

sample_data(ps)$sio2.avg <- round(na_ma(sample_data(ps)$sio2.avg, k = 1), digits = 2)

sample_data(ps)$no3.avg <- round(na_ma(sample_data(ps)$no3.avg, k = 1), digits = 2)

sample_data(ps)[is.na(sample_data(ps)$chl.a),]$chl.a <- 0
sample_data(ps)[is.na(sample_data(ps)$phaeo),]$phaeo <- 0
sample_data(ps)[is.na(sample_data(ps)$rfsu),]$rfsu <- 0

```



# Summarize sequencing depths 


```{r, message=FALSE, fig.fullwidth=TRUE}
# number of reads per samples
reads <- as.data.frame(sample_sums(ps))
colnames(reads) <- c("total")
reads$run <- sample_data(ps)$run
reads$sample <- rownames(reads)

gghistogram(reads, x = "total",
            add = "mean", rug = TRUE,
            bins = 50,
            color = MediumGrey, fill = MediumGrey,
            palette = Palette1) +
  geom_vline(xintercept = 2000, linetype = 2, col = DarkGrey) +
  clean_theme +
  xlab("Reads") + ylab("Samples") +
  labs(title = "Distribution of reads", 
       subtitle = "All samples from all 11 MiSeq runs", 
       caption = paste0("MiSeq runs (n = ", length(unique(reads$run)), ")\n",
                        "Samples (n = ", dim(reads)[1], ")"))
```

Sequencing depth for each run

```{r, fig.fullwidth=TRUE}
gghistogram(reads, x = "total",
            add = "mean", rug = TRUE,
            color = "run", fill = "run",
            bins = 50,
            #color = MediumGrey, fill = MediumGrey,
            palette = Palette1[-c(11,12)]) + 
  facet_wrap(~run) +
  geom_vline(xintercept = 2000, linetype = 2, col = DarkGrey) +
  clean_theme + theme(legend.position = "none") +
  xlab("Reads") + ylab("Samples") +
  labs(title = "Distribution of reads", 
       subtitle = "Samples amplified with EMP primers for each MiSeq run", 
       caption = paste0("MiSeq runs (n = ", length(unique(reads$run)), ")\n",
                        "Samples (n = ", dim(reads)[1], ")"))
```

```{r}
# based on the plots above define the minimum number of reads per sample
min.reads <- 2000
smp.keeper <- reads[reads$total >= min.reads,]$sample

ps0 <- ps %>%
  subset_samples(rownames(sample_data(ps)) %in% smp.keeper)
ps0 
```


# Filter taxa

EMP primers are meant to amplify prokaryotes and now we want to focus on Bacteria only.

```{r}
# because these were assigned with Silva
ps0 <- subset_taxa(ps0, Kingdom %in% c("Bacteria"))
```

## Prevalence filtering

Define prevalence of each taxa (in how many samples did each taxa appear at least once).


```{r, fig.fullwidth=TRUE}
#Define prevalence of each taxa
prev0 = apply(X = otu_table(ps0),
              MARGIN = ifelse(taxa_are_rows(ps0), yes = 1, no = 2),
              FUN = function(x){sum(x > 0)})
prevdf = data.frame(Prevalence = prev0,
                    TotalAbundance = taxa_sums(ps0),
                    tax_table(ps0))
keepPhyla = table(prevdf$Phylum)[(table(prevdf$Phylum) > 5)]
prevdf1 = subset(prevdf, Phylum %in% names(keepPhyla))

# Define prevalence threshold as 5% of total samples
prevalenceThreshold = 0.05 * nsamples(ps0)
prevalenceThreshold

# Execute prevalence filter, using `prune_taxa()` function
ps1 = prune_taxa((prev0 > prevalenceThreshold), ps0)
ps1

# Filter entries with unidentified Phylum.
ps2 = subset_taxa(ps1, Phylum %in% names(keepPhyla))
ps2

ggplot(prevdf1, aes(TotalAbundance, Prevalence, color = Phylum)) +
  geom_hline(yintercept = prevalenceThreshold, alpha = 0.5, linetype = 2) +
  geom_point(size = 2, alpha = 0.7) +
  scale_y_log10() + scale_x_log10() +
  xlab("Total Abundance") +
  facet_wrap(~Phylum) + theme_pubr() +
  #scale_color_manual(values = phylum.color) +
  tax_color_scale(ps0, "Phylum") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```


## Agglomerate taxa at the Genus level

```{r}
taxGlomRank = "Genus"
length(get_taxa_unique(ps2, taxonomic.rank = taxGlomRank))

ps3 = tax_glom(ps2, taxrank = taxGlomRank)
```

## Filter specific samples

Remove samples that were not sampled during specific cruises and samples collected during cruise B8-2010 as they were processed differently than samples collected later. 

```{r}
ps4 <- subset_samples(ps3, depth =="0" & is.na(cruise) == FALSE & cruise != "B8-2010")
ps4
```

## Focus on Gammaproteobacteria

```{r}
ps5 <- subset_taxa(ps4, Class %in% c("Gammaproteobacteria"))
ps5
```


# Normalisation

```{r}
# Normalize number of reads in each sample using median sequencing depth.
total = median(sample_sums(ps5))
standf = function(x, t=total) round(t * (x / sum(x)))
ps6 = transform_sample_counts(ps5, standf)
```


# Bar plots

```{r, fig.fullwidth=TRUE}
plot_bar2(ps6, x="smp.num", fill = "Order", y = "Abundance") +
  facet_grid(~year)
```

# Heatmap

To get a readable heatmap, one can only take ASVs that represent at least 20% of reads in at least one sample:

```{r, fig.fullwidth=TRUE}
ps7 <- filter_taxa(ps6, function(x) sum(x > total*0.20) > 0, TRUE)
ps7
otu_table(ps7)[1:8, 1:5]

plot_heatmap(ps7, method = "MDS", distance = "(A+B-2*J)/(A+B-J)",
             taxa.label = "Order", taxa.order = "Order",
             trans=NULL, low="white", high=Jeans, na.value="white") +
  clean_theme +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.position="right")
```

## Other distances

```{r}
dist_methods <- unlist(distanceMethodList)
print(dist_methods)

# A+B-2*J “quadratic” squared Euclidean
# A+B-2*J “minimum” Manhattan
# (A+B-2*J)/(A+B) “minimum” Bray-Curtis
# (A+B-2*J)/(A+B) “binary” Sørensen
# (A+B-2*J)/(A+B-J) “binary” Jaccard
```


```{r, fig.fullwidth=TRUE}
plot_richness(ps5, x = "cruise", color = "transect", measures=c("Chao1", "Shannon")) +
  clean_theme + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.position="right")
```


# Ordination

```{r, fig.fullwidth=TRUE}

ps.gamma <- subset_samples(ps6, season == "Spring")

plot_bar2(ps.gamma, x = "smp.num", y = "Abundance", fill = "Order")

ps.gamma.ord <- ordinate(ps.gamma, "NMDS", "bray")

plot_ordination(ps.gamma, 
                ps.gamma.ord, 
                type="taxa", color="Order",
                title="ASVs") +
  tax_color_scale(ps.gamma, "Order") + clean_theme + 
  theme(aspect.ratio = 1,
        legend.position = "right")
```

```{r, fig.fullwidth=TRUE}
plot_ordination(ps.gamma, ps.gamma.ord, "samples", color="region",  label = "transect") + 
  facet_wrap(~ cruise) +
  geom_point(size=7, alpha=0.2) + geom_path() + 
  guides(color=guide_legend(ncol=1)) +
  scale_color_manual(values = c(Jeans, Grapefruit)) +
  clean_theme + theme(legend.position = "right", 
                      aspect.ratio = 1)
```


# Iceland Map

```{r}
library(marmap)

xlim <- c(-30, -10)
ylim <- c(70, 62)

# get bathymetry data from NOAA
depth <- getNOAA.bathy(lon1 = xlim[1], lon2 = xlim[2],
                       lat1 = ylim[1], lat2 = ylim[2],
                       resolution = 1)

# turn the object into a data.frame
df.depth <- fortify(depth)
iceland <- map_data("world", region = "Iceland")

# select specific taxonomic group
ps.f <- subset_taxa(ps4, Family %in% c("Cyanobiaceae"))

# Normalize number of reads in each sample using median sequencing depth.
total = median(sample_sums(ps.f))
standf = function(x, t=total) round(t * (x / sum(x)))
ps.f = transform_sample_counts(ps.f, standf)

df.f <- as.data.frame(otu_table(ps.f))
head(df.f)

df.f$sample <-rownames(df.f)
df.f$stn <- substr(rownames(df.f), 1,5)

df.f$lat <- sample_data(ps.f)[sample_data(ps.f)$stn == df.f$stn,]$lat
df.f$lon <- sample_data(ps.f)[sample_data(ps.f)$stn == df.f$stn,]$lon
df.f$cruise <- sample_data(ps.f)[sample_data(ps.f)$stn == df.f$stn,]$cruise

mdf.f <- melt(df.f, id=c("sample", "stn","lat", "lon", "cruise"))
head(mdf.f)

mdf.f$cruise <- factor(mdf.f$cruise, levels =c("B8-2010", "B4-2011", "B5-2012", "B3-2013", "B4-2014", "B4-2015","B9-2016", "B11-2017", "B3-2018", "B7-2018"))

ggplot() +
  theme_bw() +
  geom_contour(data = depth, aes(x, y, z = z),
               breaks=c(-25, -50, -100, -200, -400),
               colour="black", size=0.1) +
  geom_path(data = iceland, aes(long, lat), size=0.1) +
  geom_point(data = mdf.f, aes(x = lon,
                                  y = lat,
                                  size = value, color = variable),
             alpha = 0.5) +
  facet_grid(variable~cruise) + theme(aspect.ratio=1) +
  # facet_grid(~cruise) + theme(aspect.ratio=1) +
  coord_quickmap(xlim = xlim, ylim = ylim, expand = FALSE) +
  labs(x = NULL, y = NULL)

# look the sequence
as.character(refseq(ps)["ASV2"])
```

# Correlation matrix

```{r, fig.fullwidth=TRUE}
library(phylosmith)

ps.rel <- relative_abundance(ps4)

com.tax <- common_taxa(ps.rel, treatment = NULL, subset = NULL, n = 'all')
tax_table(ps.rel)[com.tax]

com.tax.north <- common_taxa(ps.rel, treatment = "region", subset = "North", n = 'all')
tax_table(ps.rel)[com.tax.north]

com.tax.south <- common_taxa(ps.rel, treatment = "region", subset = "South", n = 'all')
tax_table(ps.rel)[com.tax.south]

taxa_core(ps.rel, treatment = NULL, subset = NULL, frequency = 0.5, abundance_threshold = 0.01)

variable_correlation_heatmap(ps.rel, variables = c('salt.avg', 'temp.avg'),
 treatment = c('region'),
 classification = 'Genus', method = 'spearman', cores = 1,
 significance_color = 'black')

variable_correlation_heatmap(ps.rel, variables = c('salt.avg', 'temp.avg'),
 treatment = c('cruise'),
 classification = 'Phylum', method = 'spearman', cores = 4,
 significance_color = 'black') + clean_theme

```


```{r}
taxa_core_graph(ps.rel, abundance_thresholds = seq(0.01, 0.25, 0.01))
```


```{r}
alpha_diversity_graph(ps.rel, index = 'shannon', treatment = c('region', 'season'), subset = NULL, colors = 'default')
```

```{r}
nmds_phyloseq(ps.rel, c("season"), circle = TRUE, colors = 'default', verbose = TRUE)
```


```{r}
pcoa_phyloseq(ps.rel, c("region"), x = 1, y = 2, method = 'bray', circle = 0.95, colors = 'default', labels = NULL)
```

```{r}
tsne_phyloseq(ps.rel, c("region"), perplexity = 10, circle = TRUE,  colors = 'default')
```


```{r}
co_occurrence_network(ps.rel, treatment = c("region", "season"), subset = "Spring", co_occurrence_table = NULL, classification = "Order", node_colors = 'default', cluster = FALSE, cluster_colors = 'default', buffer = 0.5)
```


```{r}
variable_correlation_network(ps.rel, "salt.avg", treatment = c("region", "season"), subset = "Spring", classification = "Genus")
```


```{r}
library(microbiomeSeq)
library(DESeq2)

set.seed(TRUE)
physeq <-normalise_data(ps3,norm.method = "randomsubsample")
physeq <- normalise_data(physeq, norm.method = "varstab" ,fitType="local")


physeq <- normalise_data(physeq, norm.method = "relative", norm.meta=T)

physeq <- normalise_data(ps3, norm.method = "scale", type="log")
physeq <- normalise_data(ps3, norm.method = "scale", type="sqrt")
```


```{r}
p <- plot_anova_diversity(physeq, method = c("richness","simpson", "shannon"), grouping_column = c("zone"), pValueCutoff=0.05)
print(p)
```


```{r}
p<-plot_anova_diversity(physeq, method = c("richness","simpson", "shannon"),grouping_column =  "region",pValueCutoff=0.05)
print(p)
```



```{r}
ord.res <- ordination(physeq,  method="NMDS",grouping_column="zone",pvalue.cutoff=0.05)
```

